### PR TITLE
Display name updates shortform title

### PR DESCRIPTION
**Pull request by oetherington**
User display names are denormalized into their shortform/quick takes post title. This title is not currently updated when changing the display name. This PR adds a callback to automatically update the post title, and a manual migration to resync all shortform post titles with the relevant user display names.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207789301237970) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by oetherington** [comment]
CC @b0b3rt - I've not forum gated this. I noticed LW has [some users](https://www.lesswrong.com/posts/QZM6pErzL7JwE3pkv/shortplav) who've customised the name of their quick takes posts, so maybe you want this forum gated?

Also, the logic that decides whether to name the post "quick takes" or "shortform" is still gated only to EA forum - maybe you want to change this now you're also using "quick takes"?

**Comment by b0b3rt** [comment]
@oetherington yeah I think given the existing customization of shortform titles on LessWrong, we should probably forum-gate it to avoid users clobbering custom titles.  (Another option is to check whether the previous title was the "default" title given a user's previous display name, and if so, update only in that case.  That'd also be fine for LW without forum-gating.)

I do remember at least explicitly considering the question of whether to change "shortform" to "quick take" on the LW side when implementing the rest of quick takes, so even though I don't remember the exact reasoning I'm going to go ahead and say let's stick with what's there now for LW.

File: packages/lesswrong/server/callbacks/userCallbacks.tsx [hunk 1]
```diff
@@ -464,6 +464,14 @@ getCollectionHooks("Users").updateBefore.add(async function UpdateDisplayName(da
...
     }
   }
+  if (data.shortformFeedId) {
+    void updateMutator({
+      collection: Posts,
+      documentId: data.shortformFeedId,
+      set: {title: userShortformPostTitle(data)},
+      validate: false,
+    });
+  }
```
Comment (Will-Howard) [inline packages/lesswrong/server/callbacks/userCallbacks.tsx:474]:
```suggestion
      throw new Error("This display name is already taken");
    }
    if (data.shortformFeedId) {
      void updateMutator({
        collection: Posts,
        documentId: data.shortformFeedId,
        set: {title: userShortformPostTitle(data)},
        validate: false,
      });
    }
  }
```
I think this should run inside the block that checks if the display name has actually changed, so as not to make unnecessary updates when the user edits anything on their profile.

File: packages/lesswrong/server/manualMigrations/2024-07-11-updateShortformPostTitles.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        operations.push({
+          updateOne: {
+            filter: {_id: shortformFeedId},
+            update: {$set: {title: newTitle}},
+          },
+        });
+      }
+    }
+
+    await Posts.rawCollection().bulkWrite(operations);
```
Comment (jpaddison3) [inline packages/lesswrong/server/manualMigrations/2024-07-11-updateShortformPostTitles.ts:34]: This probably works without batching. I would probably have written it with batching.

File: packages/lesswrong/server/callbacks/userCallbacks.tsx (hunk 1 above)
Comment (jpaddison3) [reply packages/lesswrong/server/callbacks/userCallbacks.tsx:474]: Agreed.

**Review by jpaddison3 — APPROVED** [review]
ASTLAC

**Pull request merged**